theano/tensor/nnet/nnet.py

```diff
@@ -465,119 +503,138 @@ def c_code_template(dtype):
+                    {
+                        dtype_%(sm)s row_ij = x_i[j * stride_x];
+                        row_max = (row_ij > row_max) ? row_ij : row_max;
+                        }
+
+                    """
 
         inside_row_loop = """
-            for (j = 0; j < Nx[1]; ++j)
-            {
-                dtype_%(sm)s row_ij = x_i[j * Sx1] ;
-                //std::cout << "2 " << j << " " << row_ij << " " << row_max << "\\n";
-                dtype_%(sm)s sm_ij = exp(row_ij - row_max);
-                //std::cout << "3 " << j << " " << sm_ij << "\\n";
-                sum += sm_ij;
-                sm_i[j * Ssm1] = sm_ij;
+        // Substract the max and take the exponential
+        for (j = 0; j < num_data; ++j)
+        {
+            dtype_%(sm)s row_ij = x_i[j * stride_x];
+            dtype_%(sm)s sm_ij = exp(row_ij - row_max);
+            sum += sm_ij;
+            sm_i[j * stride_sm] = sm_ij;
             }
 
-            //cblas_dscal(x.N, 1.0 / sum, &mat_at(s,i,0), s.n);
-            double sum_inv = 1.0 / sum;
-            for (j = 0; j < Nx[1]; ++j)
-            {
-                sm_i[j * Ssm1] *= sum_inv;
+        // cblas_dscal(x.N, 1.0 / sum, &mat_at(s,i,0), s.n);
+        // Then divide by the sum of the elements
+        double sum_inv = 1.0 / sum;
+        for (j = 0; j < num_data; ++j)
+        {
+            sm_i[j * stride_sm] *= sum_inv;
+            // std::cout << "Col: " << sm_i[j];
             }
-
         """
         # Get the vectorized version of exp if it exist
         try:
-            vec_exp = theano.scalar.exp.c_code_contiguous_raw(dtype,
-                                                              "Nx[1]", "sm_i", "sm_i")
+            vec_exp = theano.scalar.exp.c_code_contiguous_raw(dtype, "num_data", "sm_i", "sm_i")
             inside_row_loop_contig = """
-            for (j = 0; j < Nx[1]; ++j)
+            // Substract the max
+            for (j = 0; j < num_data; ++j)
             {
-                sm_i[j * Ssm1] = x_i[j * Sx1] - row_max;
-            }
+                sm_i[j * stride_sm] = x_i[j * stride_x] - row_max;
+                }
+
+            // Take the exponential
             %(vec_exp)s;
-            for (j = 0; j < Nx[1]; ++j)
+
+            // Compute the sum of exponentials
+            for (j = 0; j < num_data; ++j)
             {
-                sum += sm_i[j * Ssm1];
-            }
+                sum += sm_i[j * stride_sm];
+                }
 
+            // Then normalize each element
             //cblas_dscal(x.N, 1.0 / sum, &mat_at(s,i,0), s.n);
             double sum_inv = 1.0 / sum;
-            for (j = 0; j < Nx[1]; ++j)
+            for (j = 0; j < num_data; ++j)
             {
-                sm_i[j * Ssm1] *= sum_inv;
-            }
-
+                sm_i[j * stride_sm] *= sum_inv;
+                }
             """ % locals()
 
             inside_row_loop = """
-            if(Ssm1 == 1){
                 %(inside_row_loop_contig)s
```

Comment:
What if it's not contiguous? Shouldn't you keep the else here? Or I'm missing where are you using
```inside_row_loop```?